test/e2e/certman_operator_tests.go

```diff
@@ -62,4 +80,186 @@ var _ = Describe("Certman Operator", Ordered, func() {
 			return apiserver.Spec.ServingCerts.NamedCertificates[0].ServingCertificate.Name == secretName
 		}, pollingDuration, 30*time.Second).Should(BeTrue(), "Certificate secret should be applied to apiserver object")
 	})
+
+	It("remove finalizer from the ClusterDeployment", func(ctx context.Context) {
+		clusterDeploymentGVR := schema.GroupVersionResource{
+			Group:    "hive.openshift.io",
+			Version:  "v1",
+			Resource: "clusterdeployments",
+		}
+
+		Eventually(func() bool {
+			cdList, err := dynamicClient.Resource(clusterDeploymentGVR).Namespace("certman-operator").List(ctx, metav1.ListOptions{})
+			if err != nil || len(cdList.Items) == 0 {
+				Fail("Error listing ClusterDeployments or no ClusterDeployment found")
+			}
+
+			clusterDeployment := cdList.Items[0]
+			cdName := clusterDeployment.GetName()
+			logger.Info("Processing ClusterDeployment", "name", cdName)
+
+			clusterDeployment.SetFinalizers([]string{})
+			_, err = dynamicClient.Resource(clusterDeploymentGVR).Namespace("certman-operator").Update(ctx, &clusterDeployment, metav1.UpdateOptions{})
+			if err != nil {
+				logger.Error(err, "Failed to remove finalizers from ClusterDeployment", "name", cdName)
+				return false
+			}
+
+			logger.Info("Finalizers removed from ClusterDeployment", "name", cdName)
+
+			updatedCD, err := dynamicClient.Resource(clusterDeploymentGVR).Namespace("certman-operator").Get(ctx, cdName, metav1.GetOptions{})
+			if err != nil {
+				logger.Info("ClusterDeployment is deleted", "name", cdName)
+				return true
+			}
+
+			finalizers := updatedCD.GetFinalizers()
+			for _, finalizer := range finalizers {
+				if finalizer == "hive.openshift.io/deprovision" || finalizer == "certificaterequests.certman.managed.openshift.io" {
+					logger.Info("Finalizer has been re-added to ClusterDeployment", "name", cdName)
+					return true
+				}
+			}
+
+			logger.Info("Finalizer not yet re-added to ClusterDeployment", "name", cdName)
+			return false
+
+		}, pollingDuration, 30*time.Second).Should(BeTrue(), "ClusterDeployment should be deleted or have finalizer re-added")
+	})
+
+	It("remove finalizer from the CertificateRequest", func(ctx context.Context) {
```

Comment:
same here, the remove finalizer to certificaterequest happens only when the certificaterequest being deleted.

The function within the test which is setting finalizer to empty string did not test the operator/controller's function.

Reply:
Same fix applied - the test now verifies the operator automatically ensures the finalizer is present when not being deleted, rather than manually removing it. This properly tests the controller's functionality.